File: Chapter09/HelpDeskVB/HelpDeskVB/HelpDeskVB.DesktopClient/Screens/EngineerDashboard.lsml.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2015 by Tim Leung, published by Apress, ISBN 978-1-4842-0767-3".

Namespace LightSwitchApplication

    Public Class EngineerDashboard

        Private Sub Engineer_Changed()
            ' Write your code here.
            Me.SetDisplayNameFromEntity(Me.Engineer)
        End Sub

        Private Sub EngineerDashboard_Saved()
            ' Write your code here.
            Me.SetDisplayNameFromEntity(Me.Engineer)
        End Sub

        'Listing 9 - 1.Constructing custom text to display on a screen
        Private Sub EngineerDashboard_InitializeDataWorkspace(saveChangesTo As List(Of IDataService))
            ' Write your code here.

            For Each issue In Top5OldestIssues
                IssuesOverdueLabel += issue.Subject + Environment.NewLine
            Next


            'Listing 9 - 13.Referencing a control using ControlAvailable 

            AddHandler Me.FindControl("IssuesOverdueLabel").ControlAvailable,
                Sub(sender As Object, e As ControlAvailableEventArgs)
                    Dim issueLabel = CType(e.Control,
                    System.Windows.Controls.TextBlock)
                    issueLabel.Foreground = New SolidColorBrush(Colors.Red)
                End Sub


        End Sub


        'Listing 9-2. Setting the screen title in code
        Private Sub Engineer_Loaded(succeeded As Boolean)
            ' Write your code here.
            Me.DisplayName = "Engineer Dashboard"

        End Sub

        'Listing 9-5. Opening screens from an unbound drop-down

        Private Sub OpenIssue_Execute()
            Me.Application.ShowIssueDetail(IssueSelectionProperty.Id)
        End Sub

        Private Sub RefreshAllScreens_Execute()

            'Listing 9 - 6.Refreshing all open screens
            Dim screens = Me.Application.ActiveScreens
            For Each s In screens
                Dim screen = s.Screen
                screen.Details.Dispatcher.BeginInvoke(
         Sub()
             screen.Refresh()
         End Sub)
            Next


        End Sub

        Private Sub ArchiveIssues_Execute()
            ' Write your code here.
            If Me.ShowMessageBox(
            "Are you sure you want delete all issues older than 12 months? (This is a demonstration of how to use a MessageBox)",
            "Confirm Delete", MessageBoxOption.YesNo) =
            System.Windows.MessageBoxResult.Yes Then
                'DeleteOldIssues()
            End If

        End Sub

        'Listing 9-20. Opening the combination Add/Edit screen to add a New record
        Private Sub OpenNewIssueScreen_Execute()
            Application.ShowAddEditIssue(Nothing)
        End Sub

    End Class

End Namespace

File: Chapter09/HelpDeskCS/HelpDeskCS/HelpDeskCS.DesktopClient/Screens/EngineerDashboard.lsml.cs
// You can use and redistribute the code from this book (and sample application) for non-commercial and 
// most commercial purposes as long as you acknowledge the source and authorship. 
// You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
// The acknowledgment should include author, title, publisher, and ISBN. 
// An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2015 by Tim Leung, published by Apress, ISBN 978-1-4842-0767-3".

using Microsoft.LightSwitch.Presentation.Extensions;
using Microsoft.LightSwitch.Presentation;
using Microsoft.LightSwitch.Framework.Client;
using Microsoft.LightSwitch;
using System.Collections.Generic;
using System.IO.IsolatedStorage;
using System.IO;
using System.Linq;
using System;
using System.Windows.Media;

namespace LightSwitchApplication
{
    public partial class EngineerDashboard
    {
        partial void Engineer_Loaded(bool succeeded)
        {



            //Listing 9-2. Setting the screen title in code
            this.DisplayName = "Engineer Dashboard";

        }

        partial void Engineer_Changed()
        {
            // Write your code here.
            this.SetDisplayNameFromEntity(this.Engineer);
        }

        partial void EngineerDashboard_Saved()
        {
            // Write your code here.
            this.SetDisplayNameFromEntity(this.Engineer);
        }

        partial void EngineerDashboard_InitializeDataWorkspace(List<IDataService> saveChangesTo)
        {
            // Write your code here.


            foreach (Issue issue in Top5OldestIssues)
            {
                IssuesOverdueLabel += issue.Subject + Environment.NewLine;
            }


            //Listing 9 - 13.Referencing a control using ControlAvailable

            var control = this.FindControl("IssuesOverdueLabel");

            control.ControlAvailable +=
            (object sender, ControlAvailableEventArgs e) =>
            {
                var issueLabel =
                    (System.Windows.Controls.TextBlock)e.Control;
                issueLabel.Foreground = new SolidColorBrush(Colors.Red);
            };

        }

        partial void OpenIssue_Execute()
        {


            //Listing 9-5. Opening screens from an unbound drop-down
            this.Application.ShowIssueDetail(IssueSelectionProperty.Id);


        }

        partial void RefreshAllScreens_Execute()
        {
            //Listing 9 - 6.Refreshing all open screens

            var screens = this.Application.ActiveScreens;
            foreach (var s in screens)
            {
                var screen = s.Screen;
                screen.Details.Dispatcher.BeginInvoke(() =>
                {
                    screen.Refresh();
                });
            }

        }

        partial void ArchiveIssues_Execute()
        {
            //Listing 9 - 9.Displaying a message box

            if (this.ShowMessageBox(
            "Are you sure you want delete all issues older than 12 months? (This is a demonstration of how to use a MessageBox)",
                "Confirm Delete", MessageBoxOption.YesNo) ==
                System.Windows.MessageBoxResult.Yes)
            {
                //DeleteOldIssues();
            }
        }

        partial void OpenNewIssueScreen_Execute()
        {

            //Listing 9-20. Opening the combination Add/Edit screen to add a New record
            this.Application.ShowAddEditIssue(null);

        }

        partial void CreateNewIssue_Execute()
        {
            // Write your code here.
            this.Application.ShowCreateNewIssue(this.Engineer.Id);
        }
    }
}
